src/librustc_trans/mir/block.rs

```diff
@@ -467,34 +461,28 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
                     let drop_fn = glue::get_drop_glue(bcx.ccx, ty);
                     let drop_ty = glue::get_drop_glue_type(bcx.ccx.shared(), ty);
                     let is_sized = bcx.ccx.shared().type_is_sized(ty);
-                    let llvalue = if is_sized {
-                        if drop_ty != ty {
-                            bcx.pointercast(llval, type_of::type_of(bcx.ccx, drop_ty).ptr_to())
+                    let (sized_args, unsized_args);
+                    let args: &[ValueRef] = if is_sized {
+                        sized_args = if drop_ty != ty {
+                            [bcx.pointercast(llval, type_of::type_of(bcx.ccx, drop_ty).ptr_to())]
                         } else {
-                            llval
-                        }
+                            [llval]
+                        };
+                        &sized_args
                     } else {
-                        // FIXME(#36457) Currently drop glue takes sized
-                        // values as a `*(data, meta)`, but elsewhere in
-                        // MIR we pass `(data, meta)` as two separate
-                        // arguments. It would be better to fix drop glue,
-                        // but I am shooting for a quick fix to #35546
-                        // here that can be cleanly backported to beta, so
-                        // I want to avoid touching all of trans.
-                        let scratch = base::alloc_ty(&bcx, ty, "drop");
-                        Lifetime::Start.call(&bcx, scratch);
-                        bcx.store(llval, base::get_dataptr(&bcx, scratch));
-                        bcx.store(llextra, base::get_meta(&bcx, scratch));
-                        scratch
+                        unsized_args = [llval, llextra];
+                        &unsized_args
                     };
                     if let Some(unwind) = *cleanup {
-                        bcx.invoke(drop_fn,
-                            &[llvalue],
+                        bcx.invoke(
+                            drop_fn,
+                            args,
                             self.blocks[target],
                             llblock(self, unwind),
-                            cleanup_bundle);
+                            cleanup_bundle
+                        );
                     } else {
-                        bcx.call(drop_fn, &[llvalue], cleanup_bundle);
+                        bcx.call(drop_fn, args, cleanup_bundle);
```

Comment:
Here's the cool part: since the ABI is identical for the drop glue and `drop_in_place`, you can just change `callee.data` to `Fn(drop_fn)` and mutate `intrinsic` to `None`. You can replace the pointer cast of an argument with a cast of `drop_fn` to the LLVM type of `fn_ty` (which means the `if` has to be moved lower down to where `fn_ty` is declared).